slurm inclusion

replikation opened the pull request:
* for testing slurm @hoelzer @mshamash
* This branch includes already the latest "test" profile which supplies test data (just add it to the `-profile`)

```bash
git clone https://github.com/replikation/What_the_Phage.git && cd What_the_Phage
git checkout slurm_exec
./phage.nf -profile test,slurm,docker
# or
./phage.nf -profile test,slurm,singularity
```

if something goes wrong please paste the full error terminal print in here.

mshamash commented:
Thanks for making the updated branch for Slurm compatibility! Looks like our HPC cluster is using Nextflow 19.10.0 and the WtP workflow requires 20.X ++. I'll have to reach out to the team who manages this to see if they can install 20.X, but this will delay my testing by at least a week.

`$ ./What_the_Phage/phage.nf -profile test,slurm,docker
N E X T F L O W  ~  version 19.10.0
Launching `./What_the_Phage/phage.nf` [jovial_mestorf] - revision: b33954ed63
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
This workflow requires Nextflow version 20.X or greater -- You are running version 19.10.0`

hoelzer commented:
> Thanks for making the updated branch for Slurm compatibility! Looks like our HPC cluster is using Nextflow 19.10.0 and the WtP workflow requires 20.X ++. I'll have to reach out to the team who manages this to see if they can install 20.X, but this will delay my testing by at least a week.
> 
> `$ ./What_the_Phage/phage.nf -profile test,slurm,singularity`
> `N E X T F L O W ~ version 19.10.0`
> `Launching "./What_the_Phage/phage.nf" [jovial_mestorf] - revision: b33954ed63`
> `WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE`
> `This workflow requires Nextflow version 20.X or greater -- You are running version 19.10.0`

Maybe you can install the newest NF version in a conda environment and use this for testing? 

```bash
conda create -n nextflow -c bioconda nextflow
```

mshamash commented:
> Maybe you can install the newest NF version in a conda environment and use this for testing?
> 
> ```shell
> conda create -n nextflow -c bioconda nextflow
> conda activate nextflow
> nextflow run ...
> ```

No conda on our HPC either 😬 Good point though, I will look into installing NF in a python venv, if this can be done, since we're allowed venvs on this HPC.

replikation commented:
@mshamash  I think you don't need to "install nextflow" just do a "curl -s https://get.nextflow.io | bash" and you get an executable file in your current working dir. you can use this one. you only need admin rights if you want to add it to the PATH. at least this was my understanding of it.

or download the executable directly from here: https://github.com/nextflow-io/nextflow/releases/download/v20.04.1/nextflow
i think nextflow downloads all dependencies directly to your current workdir of the executable.

mshamash commented:
@replikation thanks that worked, I had actually just tried it before you sent your message.

Got WtP to start running, and then an error occurred

`$ ./What_the_Phage/phage.nf -profile test,slurm,docker
N E X T F L O W  ~  version 20.04.1
Launching "./What_the_Phage/phage.nf" [reverent_ardinghelli] - revision: b33954ed63
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
_____ _____ ____ ____ ___ ___ __ __ _ _ 
  __      _______________________ 
 /  \    /  \__    ___/\______   \
 \   \/\/   / |    |    |     ___/
  \        /  |    |    |    |    
   \__/\  /   |____|    |____|    
        \/                        
_____ _____ ____ ____ ___ ___ __ __ _ _ 
 
Profile: test,slurm,docker
 
Current User: mshamash
Nextflow-version: 20.04.1
WtP intended for Nextflow-version: 20.01.0
Starting time: 03-05-2020 19:37 UTC
Workdir location [--workdir]:
  /tmp/nextflow-phages-mshamash
Output location [--output]:
  results
Database location [--databases]:
  nextflow-autodownload-databases
 
CPUs to use: 8
 
[-        ] process > get_test_data:testprofile                                               -
[-        ] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (1)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (2)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (3)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (4)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (5)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (6)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (8)
[d5/ca1c83] process > get_test_data:testprofile                                               [  0%] 0 of 1
executor >  slurm (9)
[d5/ca1c83] process > get_test_data:testprofile                                               [100%] 1 of 1, failed: 1 ✘
[25/040303] process > phage_references:download_references                                    [100%] 1 of 1, failed: 1 ✘
[-        ] process > phage_blast_DB:phage_references_blastDB                                 -
[-        ] process > ppr_dependecies:ppr_download_dependencies                               -
[-        ] process > vibrant_download_DB                                                     -
[-        ] process > virsorter_database:virsorter_download_DB                                -
[-        ] process > pvog_database:pvog_DB                                                   -
[-        ] process > vog_database:vog_DB                                                     -
[-        ] process > rvdb_database:rvdb_DB                                                   -
[-        ] process > checkV_database:download_checkV_DB                                      -
[-        ] process > sourmash_database:sourmash_download_DB                                  -
[-        ] process > identify_fasta_MSF:fasta_validation_wf:input_suffix_check               -
[-        ] process > identify_fasta_MSF:fasta_validation_wf:seqkit                           -
[-        ] process > identify_fasta_MSF:virsorter_wf:virsorter                               -
[-        ] process > identify_fasta_MSF:virsorter_wf:filter_virsorter                        -
[-        ] process > identify_fasta_MSF:virsorter_wf:virsorter_collect_data                  -
[-        ] process > identify_fasta_MSF:marvel_wf:split_multi_fasta                          -
[-        ] process > identify_fasta_MSF:marvel_wf:marvel                                     -
[-        ] process > identify_fasta_MSF:marvel_wf:filter_marvel                              -
[-        ] process > identify_fasta_MSF:marvel_wf:marvel_collect_data                        -
[-        ] process > identify_fasta_MSF:sourmash_wf:split_multi_fasta                        -
[-        ] process > identify_fasta_MSF:sourmash_wf:sourmash                                 -
[-        ] process > identify_fasta_MSF:sourmash_wf:filter_sourmash                          -
[-        ] process > identify_fasta_MSF:sourmash_wf:sourmash_collect_data                    -
[-        ] process > identify_fasta_MSF:metaphinder_wf:metaphinder                           -
[-        ] process > identify_fasta_MSF:metaphinder_wf:filter_metaphinder                    -
[-        ] process > identify_fasta_MSF:metaphinder_wf:metaphinder_collect_data              -
[-        ] process > identify_fasta_MSF:metaphinder_own_DB_wf:metaphinder_own_DB             -
[-        ] process > identify_fasta_MSF:metaphinder_own_DB_wf:filter_metaphinder_own_DB      -
[-        ] process > identify_fasta_MSF:metaphinder_own_DB_wf:metaphinder_collect_data_ownDB -
[-        ] process > identify_fasta_MSF:deepvirfinder_wf:deepvirfinder                       -
[-        ] process > identify_fasta_MSF:deepvirfinder_wf:filter_deepvirfinder                -
[-        ] process > identify_fasta_MSF:deepvirfinder_wf:deepvirfinder_collect_data          -
[-        ] process > identify_fasta_MSF:virfinder_wf:virfinder                               -
[-        ] process > identify_fasta_MSF:virfinder_wf:filter_virfinder                        -
[-        ] process > identify_fasta_MSF:virfinder_wf:virfinder_collect_data                  -
[-        ] process > identify_fasta_MSF:pprmeta_wf:pprmeta                                   -
[-        ] process > identify_fasta_MSF:pprmeta_wf:filter_PPRmeta                            -
[-        ] process > identify_fasta_MSF:pprmeta_wf:pprmeta_collect_data                      -
[-        ] process > identify_fasta_MSF:vibrant_wf:vibrant                                   -
[-        ] process > identify_fasta_MSF:vibrant_wf:filter_vibrant                            -
[-        ] process > identify_fasta_MSF:vibrant_wf:vibrant_collect_data                      -
[-        ] process > identify_fasta_MSF:virnet_wf:normalize_contig_size                      -
[-        ] process > identify_fasta_MSF:virnet_wf:virnet                                     -
[-        ] process > identify_fasta_MSF:virnet_wf:filter_virnet                              -
[-        ] process > identify_fasta_MSF:virnet_wf:virnet_collect_data                        -
[-        ] process > identify_fasta_MSF:filter_tool_names                                    -
[-        ] process > identify_fasta_MSF:r_plot                                               -
[-        ] process > identify_fasta_MSF:upsetr_plot                                          -
[-        ] process > identify_fasta_MSF:samtools                                             -
[-        ] process > phage_annotation_MSF:prodigal                                           -
[-        ] process > phage_annotation_MSF:hmmscan                                            -
[-        ] process > phage_annotation_MSF:chromomap_parser                                   -
[-        ] process > phage_annotation_MSF:chromomap                                          -
[-        ] process > checkV_wf:checkV                                                        -
[-        ] process > phage_tax_classification:split_multi_fasta                              -
[-        ] process > phage_tax_classification:sourmash_for_tax                               -
Oops .. something went wrong
WARN: Killing pending tasks (8)
Error executing process > 'get_test_data:testprofile'

Caused by:
  Process "get_test_data:testprofile" terminated for an unknown reason -- Likely it has been terminated by the external system

Command executed:

  git clone https://github.com/mult1fractal/WtP_test-data.git

Command exit status:
  -

Command output:
  (empty)

Work dir:
  /tmp/nextflow-phages-mshamash/d5/ca1c838cdbec32378f4cdafcb2d7c8

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named ".command.sh"`

mshamash commented:
Managed to fix OOM errors by setting default job memory to 4GB unless otherwise specified in node.config file. Now having trouble with Singularity on HPC it seems. I'm not that familiar with singularity, so I'll paste the error below.

```
Oops .. something went wrong
Waiting files transfer to complete (4 files)
WARN: Killing pending tasks (6)
Error executing process > 'sourmash_database:sourmash_download_DB'

Caused by:
  Failed to pull singularity image
  command: singularity pull  --name nanozoo-sourmash-2.3.0--4257650.img docker://nanozoo/sourmash:2.3.0--4257650 > /dev/null
  status : 255
  message:
    time="2020-07-15T08:57:44-07:00" level=warning msg="\"/run/user/3087957\" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: 
Trying to pull image in the event that it is a public image."
    INFO:    Converting OCI blobs to SIF format
    INFO:    Starting build...
    time="2020-07-15T08:57:46-07:00" level=warning msg="\"/run/user/3087957\" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: 
Trying to pull image in the event that it is a public image."
    time="2020-07-15T08:57:47-07:00" level=warning msg="\"/run/user/3087957\" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: 
Trying to pull image in the event that it is a public image."
    Getting image source signatures
    Copying blob sha256:b8f262c62ec67f02536f49654de586c022043652bbb6bbf76a8dab1542627a8d
    Copying blob sha256:0a43c0154f168ca6fe36f31e366dc85ba2f95da95a9bf69399d927d513d501f9
    Copying blob sha256:906d7b5da8fb08f9d6c98d3e7df1d621d03d249c26524b64b09641b1f70fc27e
    Copying blob sha256:996f151465c0236204d313b28975d81a1685e9a75fa31bcf7d725a2f08873c8f
    Copying blob sha256:9ae3db39e36a27aa1c9fb53305313f189c5db6a257e186362c3bd40fd42ca299
    Copying blob sha256:60a5a6198fb7a49f7870d170d7f008fa1a76e3ec3cccca8577542c757c215199
    Copying config sha256:147d6d52778bb5746c4d4d2615ed6301aab150db23c5110805d750b1c44329f0
    Writing manifest to image destination
    Storing signatures
    FATAL:   While making image from oci registry: while building SIF from layers: conveyor failed to get: no descriptor found for reference "d5148042df237990b440b1fa49b9548a64e57c659e8d759946b8fb19bbc19af8"
```

mshamash commented:
@hoelzer 

I'm running Singularity v3.5.2, that's the latest we have on our HPC.

When I run that pull command, I can successfully generate the image nanozoo-sourmash-2.3.0--4257650.img. Output pasted below. 

```
$ singularity pull  --name nanozoo-sourmash-2.3.0--4257650.img docker://nanozoo/sourmash:2.3.0--4257650
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob b8f262c62ec6 skipped: already exists
Copying blob 0a43c0154f16 skipped: already exists
Copying blob 906d7b5da8fb skipped: already exists
Copying blob 996f151465c0 skipped: already exists
Copying blob 9ae3db39e36a skipped: already exists
Copying blob 60a5a6198fb7 skipped: already exists
Copying config 147d6d5277 done
Writing manifest to image destination
Storing signatures
2020/07/15 09:32:10  info unpack layer: sha256:b8f262c62ec67f02536f49654de586c022043652bbb6bbf76a8dab1542627a8d
2020/07/15 09:32:12  info unpack layer: sha256:0a43c0154f168ca6fe36f31e366dc85ba2f95da95a9bf69399d927d513d501f9
2020/07/15 09:32:18  info unpack layer: sha256:906d7b5da8fb08f9d6c98d3e7df1d621d03d249c26524b64b09641b1f70fc27e
2020/07/15 09:32:22  info unpack layer: sha256:996f151465c0236204d313b28975d81a1685e9a75fa31bcf7d725a2f08873c8f
2020/07/15 09:32:22  info unpack layer: sha256:9ae3db39e36a27aa1c9fb53305313f189c5db6a257e186362c3bd40fd42ca299
2020/07/15 09:32:22  info unpack layer: sha256:60a5a6198fb7a49f7870d170d7f008fa1a76e3ec3cccca8577542c757c215199
INFO:    Creating SIF file...
INFO:    Build complete: nanozoo-sourmash-2.3.0--4257650.img
```

hoelzer commented:
Okay, so NF failed to generate your image but it works in principle. 

I think the problem could be that you have to define a cache dir for your singularity images. 

The default is: 
```
cachedir = "singularity_images"
```

but you might want to point this to somewhere else where you have full read/write access and enough space. And you might also want to adjust the `workdir = "/tmp/nextflow-phages-$USER"`

Can you try

```bash
./What_the_Phage/phage.nf -profile test,slurm,singularity --workdir /some/path --cachedir /some/other/path
```

@replikation also figured out that the initial download of the singularity images sometimes fails, so you might also want to try this:
```
  WARNING: Singularity image building sometimes fails!
  Please download all images first via --setup --cachedir IMAGE-LOCATION
  Manually remove faulty images in singularity_images for a rebuild
```

replikation commented:
According to the --setup a few Images were created, i also dont know why singularity fails sometimes. But somehow it has issues on your hpc to finde the location. Try rerunning --setup it skips images that exists.or maybe a disk space issue?

mshamash commented:
I've resorted to installing all singularity images manually.

However, installing the hmmscan image seems to fail consistently. This is the last one I've got to get installed before WtP can continue its analysis...

```
$ singularity pull --name mhoelzer-hmmscan-0.1.img docker://mhoelzer/hmmscan:0.1
WARN[0000] "/run/user/3087957" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: Trying to pull image in the event that it is a public image. 
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
WARN[0001] "/run/user/3087957" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: Trying to pull image in the event that it is a public image. 
WARN[0002] "/run/user/3087957" directory set by $XDG_RUNTIME_DIR does not exist. Either create the directory or unset $XDG_RUNTIME_DIR.: stat /run/user/3087957: no such file or directory: Trying to pull image in the event that it is a public image. 
Getting image source signatures
Copying blob ab5ef0e58194 skipped: already exists
Copying blob f083460cc0e6 skipped: already exists
Copying blob 30405975171a skipped: already exists
Copying blob ccb8d06f8fad skipped: already exists
Copying blob 9afa2e612b5c skipped: already exists
Copying config db6fd1a683 done
Writing manifest to image destination
Storing signatures
2020/07/16 06:03:03  info unpack layer: sha256:ab5ef0e5819490abe86106fd9f4381123e37a03e80e650be39f7938d30ecb530
2020/07/16 06:03:04  warn rootless{usr/bin/newgidmap} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/16 06:03:04  warn rootless{usr/bin/newuidmap} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/16 06:03:04  warn rootless{usr/bin/ping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/16 06:03:07  warn rootless{usr/sbin/arping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/16 06:03:07  warn rootless{usr/sbin/clockdiff} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/16 06:03:09  info unpack layer: sha256:f083460cc0e60e91afe0644b550b10f82b0f7f5ebf6a771498784d23ef55cb8b
2020/07/16 06:03:11  info unpack layer: sha256:30405975171a3b52f024633e4f8c0eeb2319df85fadd822a9a9b713f699dc147
2020/07/16 06:03:11  warn rootless{usr/bin/ssh-agent} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:13  warn rootless{usr/libexec/openssh/ssh-keysign} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:14  info unpack layer: sha256:ccb8d06f8fad73ae8e11a63c3774872eec4bc5587f26b07344f73f60687b3d16
2020/07/16 06:03:15  info unpack layer: sha256:9afa2e612b5c016911a7d983ce9c5d79a738b90d11a05757f3b8b221736436ec
2020/07/16 06:03:15  warn rootless{vpHMM_database} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:38  warn rootless{vpHMM_database/vpHMM_database} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:48  warn rootless{vpHMM_database/vpHMM_database.h3f} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:48  warn rootless{vpHMM_database/vpHMM_database.h3i} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:03:59  warn rootless{vpHMM_database/vpHMM_database.h3m} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/16 06:04:21  warn rootless{vpHMM_database/vpHMM_database.h3p} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
INFO:    Creating SIF file...
FATAL:   While making image from oci registry: while building SIF from layers: while creating squashfs: create command failed: signal: killed: Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/etc/gshadow, creating empty file
Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/etc/gshadow-, creating empty file
Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/etc/shadow, creating empty file
Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/etc/shadow-, creating empty file
Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/usr/bin/ssh-agent, creating empty file
Failed to read file /localscratch/mshamash.46204158.0/rootfs-9598733c-c764-11ea-b428-7cd30ac72600/usr/libexec/openssh/ssh-keysign, creating empty file
```

replikation commented:
In the hhmscan container is a 2GB database... even though we don't need this one I think (@mult1fractal ?) as we are adding our own database. I will build another hmm scan docker, test this and commit the changes into this branch. I'll also test if I can build a singularity image out of this new container. ill tag you when its done. Thank you

replikation commented:
@mshamash there was a small `--setup` bug that is now fixed, I replaced the docker container for hmm scan that was causing issues in your case

I tested the following:
1.) create all singularity images locally (didn't have any errors, I suggest building/downloading singularly on a normal machine, afterwards transfer the images to you HPC  and tell WtP where they are located)
```bash
./phage.nf -profile singularity,local --setup --cachedir singularity_images
``` 

2.) used these images via
```bash
./phage.nf --fasta test-data/all_pos_phage.fa -profile singularity,local --cachedir singularity_images
```

replikation commented:
* not sure if there is some kind of singularity hub like dockerhub? would help if we could give people ready to use singularity images like the docker containers

hoelzer commented:
> * not sure if there is some kind of singularity hub like dockerhub? would help if we could give people ready to use singularity images like the docker containers

I think it is (maybe https://singularity-hub.org/) but let me ask someone

mshamash commented:
@replikation I will give that a try. I also made some edits to the /configs/node.config file for a 'default' job with more memory and longer runtime. Not sure how I can add that to this PR or if I need to make a new one from a fork I just made...

replikation commented:
you can copy the code here if you want to and ill add that

mshamash commented:
@replikation 

What_the_Phage/configs/node.config 
```
process {
            time = '2h'
            memory = '4 GB'
            withLabel: checkV ... //continue with regular file// ...
}         
```

replikation commented:
i added it to the SLURM configuration [here](https://github.com/replikation/What_the_Phage/blob/ad372f0f7dd49e408a30260805d948a2ae0e7088/nextflow.config#L77) should also work for you I hope

mshamash commented:
Thanks, that config worked . Encountering some issues with hmmscan, error is below. I'm trying to re-run everything now deleting all temp cache nextflow files as well as those in /tmp/ from WtP. On another note, I noticed when running the pipeline again it always redownloads the databases, even if they're already present in the correct nextflow-autodownload-databases directory.

Also, if I have 8 contigs, does it run each tool 8x? Couldn't tell if this was the case or not from the squeue.

```
Oops .. something went wrong
WARN: Killing pending tasks (15)
Error executing process > 'phage_annotation_MSF:hmmscan (1)'

Caused by:
  Process `phage_annotation_MSF:hmmscan (1)` terminated with an error exit status (1)

Command executed:

  hmmscan --cpu 4 --noali --domtblout ERR576943_raw_assembly.fasta_project_hmmscan.tbl project/project.hmm ERR576943_raw_assembly.fasta_prodigal.faa

Command exit status:
  1

Command output:
  (empty)

Command error:
  Error: File existence/permissions problem in trying to open HMM file project/project.hmm.
  HMM file project/project.hmm not found (nor an .h3m binary of it); also looked in PFAMDB

Work dir:
  /scratch/mshamash/nextflow-phages/f0/a843ff974817e46f5a5a31ed0321ab

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`
```

replikation commented:
* that looks like a bug with the database, ill check that but could be slurm related as we did not encounter this yet
* if you run 8 contigs (in one file) each tool runs once but there are multiple processes in general "around" each tool
* @mult1fractal can you take a look at the hmmer error something is wrong with the database
   * `project/project.hmm` doesn't seem right, did not encounter this locally
   * @mshamash what was the command used for the last run, also can you show me the file content in `/scratch/mshamash/nextflow-phages/f0/a843ff974817e46f5a5a31ed0321ab`

mshamash commented:
@replikation  here is the content, I re-ran just to make sure it was a reproduceable error, hence different folder names. Also, that last symlink doesn't exist...I'm not sure how it generated that path, but that project folder doesn't exist.

```
-rw-r----- 1 mshamash mshamash    0 Jul 17 13:26 .command.begin
-rw-r----- 1 mshamash mshamash  181 Jul 17 13:26 .command.err
-rw-r----- 1 mshamash mshamash  181 Jul 17 13:26 .command.log
-rw-r----- 1 mshamash mshamash    0 Jul 17 13:26 .command.out
-rw-r----- 1 mshamash mshamash 3619 Jul 17 13:25 .command.run
-rw-r----- 1 mshamash mshamash  163 Jul 17 13:25 .command.sh
lrwxrwxrwx 1 mshamash mshamash  109 Jul 17 13:26 ERR575692_raw_assembly.fasta_prodigal.faa -> /scratch/mshamash/nextflow-phages/f2/8cb8781594b7d821ea17f1f2eeb97a/ERR575692_raw_assembly.fasta_prodigal.faa
-rw-r----- 1 mshamash mshamash    1 Jul 17 13:26 .exitcode
lrwxrwxrwx 1 mshamash mshamash   37 Jul 17 13:26 project -> /project/6003890/mshamash/wtp/project
```

replikation commented:
* yeah this "project" should actually link to the pvog_db database 
  * I think I might now whats causing this, should be related to the pvog Db download module
  * ill get back to you when iam done, and thanks for testing it for us on slurm
* but in principle, the other steps are working on slurm?
  * e.g. if you run WtP with the `--identify` flag it skips annotation and such and collects everything into a fasta file

mshamash commented:
Thus far, it seems like most of the other steps are working. I just tried using the `--identify` flag (along with `--ma`, see below) and the final output is many folders, each entitled ___.FASTA? See attached photo. The tool does not error, though, and completed the workflow fine with these two parameters.

MARVEL seems to consistently fail, on every run, for "4 of 8" where Nextflow outputs this:
`NOTE: Process "identify_fasta_MSF:marvel_wf:marvel (8)" terminated with an error exit status (1) -- Error is ignored`

I need to look into this some more, not sure why its so consistent. Let me know if there are any specific files you would like contents for from MARVEL jobs. I saved the hashes.

Other than that, the other main issue I was having was downloading Singularity containers automatically, sometimes it worked, sometimes it failed. All disks have plenty of space available, I/O, etc. As I mentioned earlier, I resorted to downloading them all manually in the corresponding directory.

<img width="439" alt="Screen Shot 2020-07-17 at 5 29 43 PM" src="https://user-images.githubusercontent.com/8537248/87832099-34afd200-c853-11ea-9cb3-b97d4558a254.png">

replikation commented:
* that some tools fail is expected and is an issue on the tool side, so don't worry, that's also why we choose a bigger multitool approach
* that is why the workflow allows for tools to fail in the identification phase
* a good catch on the .fasta dirs, ill fix that, this should only happen if `test` is used
* DB fill will be ready in a few days
* the singularity building is annoying but I cant do much about this, e.g. building them locally works fine - but I could change the `--setup` to automatically rerun a few times for the singularity buuilding

replikation commented:
@mshamash pushed some databases fixes I hope it works now
* it did not have the time to check why its redownloading databases in your case
  * would be good to know if its all databases or just a few of them that redownload
  * i assume that it might be the DBs with a "dir" as input (and not a specific file)

mshamash commented:
@replikation thanks for the update, I will give it a try today/tomorrow. Do you recommend I delete all database files and nextflow working/temp files before trying again (to force it to rebuild/download hmmscan DB to correct location)?

Separately, I think WtP redownloads the databases because at one point I was clearing out the workdir (on HPC scratch) after each run. Even though the database files were saved in the nextflow-autodownload-databases folder, I believe this triggered the redownload. I will have to do more testing to see if it's all DBs, but it was definitely the VIBRANT database, maybe others too.

replikation commented:
yep remove the Dbs to be sure, removing the workdir should not trigger a DB
redownload. There is a "checkifexist" code in WtP, so it looks for each DB
in the --database dir. But I think there is a pathing issue or maybe only
for some, I'll see if i can reproduce that in the cloud. Locally i use the
storeDir function so this is automanaged.

Also the --setup now trys a few times a rebuilding of singularity images.
Might help.

Michael Shamash <notifications@github.com> schrieb am Mo., 20. Juli 2020,
18:53:

> @replikation <https://github.com/replikation> thanks for the update, I
> will give it a try today/tomorrow. Do you recommend I delete all database
> files and nextflow working/temp files before trying again (to force it to
> rebuild/download hmmscan DB to correct location)?
>
> Separately, I think WtP redownloads the databases because at one point I
> was clearing out the workdir (on HPC scratch) after each run. Even though
> the database files were saved in the nextflow-autodownload-databases
> folder, I believe this triggered the redownload. I will have to do more
> testing to see if it's all DBs, but it was definitely the VIBRANT database,
> maybe others too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/replikation/What_the_Phage/pull/84#issuecomment-661183610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIQSLLQYXZKSVIS7TJ4TKXLR4RZBBANCNFSM4O2RATAA>
> .
>

mshamash commented:
@replikation I get an error right when I start the script (after removing DBs and workdir)

```
[-        ] process > get_test_data:testprofile            -
[-        ] process > phage_references:download_references -
Unknown workflow parameter definition: phage_blast_DB

 -- Check script 'wtp/What_the_Phage/phage.nf' at line: 663 or see '.nextflow.log' file for more details
```

replikation commented:
* i cant reproduce the bug, what was the command that you used?

replikation commented:
@mshamash the bug is fixed now. i noticed that on an e.g. slurm,lsf executer currently 3 databases are not recognised and always redownloaded ( vibrant, checkV and the phage_blastDB) ill investigate this

replikation commented:
* DBs should be properly identified by WtP

mshamash commented:
The bug is fixed, however something new came up with VirSorter... For reference, this is the command I'm running:

```
./wtp/What_the_Phage/phage.nf --workdir /scratch/mshamash/nextflow-phages2 --database wtp-autodownload_databases --cachedir wtp-singularity_images --output wtp-results_test -profile test,slurm,singularity
```

```
Oops .. something went wrong
WARN: Killing pending tasks (17)
Error executing process > 'identify_fasta_MSF:virsorter_wf:virsorter_collect_data (3)'

Caused by:
  Process `identify_fasta_MSF:virsorter_wf:virsorter_collect_data (3)` terminated with an error exit status (1)

Command executed:

  mkdir -p virsorter
  cp virsorter_results_9671cb504d637f0041a67f27b45f34.tar virsorter
  tar -czf virsorter_results_ERR576945_raw_assembly.fasta.tar.gz virsorter

Command exit status:
  1

Command output:
  (empty)

Command error:
  tar: virsorter/virsorter_results_9671cb504d637f0041a67f27b45f34.tar: file changed as we read it

Work dir:
  /scratch/mshamash/nextflow-phages2/a4/5b6c9eb69ad271c8a1e02e5c1d5efe

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`
```

replikation commented:
* @mshamash  can you do a -resume on your command and tell me if the error occurs again?
* like this:
```
./wtp/What_the_Phage/phage.nf --workdir /scratch/mshamash/nextflow-phages2 --database wtp-autodownload_databases --cachedir wtp-singularity_images --output wtp-results_test -profile test,slurm,singularity -resume
```
* (https://stackoverflow.com/questions/20318852/tar-file-changed-as-we-read-it)  could be something from your HPC and not from WtP  
   * "This error was sporadic." but i could supress the error if this sometimes occurs on your HPC, never had this
   * thanks for reporting

mshamash commented:
@replikation running with -resume seemed to have fixed it, perhaps it was a one-time thing.

Something I notice again, when running the tools (ie. VIBRANT here), it seems to make 8 VIBRANT jobs for the 8 contigs. Each with 2hr runtime, 4 CPUs, 8G RAM, each named "nf-identify_fasta_MSF_vibrant+".

So while nextflow says this:
`[f6/52394c] process > identify_fasta_MSF:vibrant_wf:vibrant (4)                                   [ 0%] 0 of 8`

Slurm squeue says this:
```
       jobid1 mshamash def-redacted nf-identify_fa   R    1:54:38     1    4        N/A      8G node1 (None) 
       jobid2 mshamash def-redacted nf-identify_fa   R    1:54:38     1    4        N/A      8G node2 (None) 
       jobid3 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node3 (None) 
       jobid4 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node4 (None) 
       jobid5 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node5 (None) 
       jobid6 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node6 (None) 
       jobid7 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node7 (None) 
       jobid8 mshamash def-redacted nf-identify_fa   R    1:54:46     1    4        N/A      8G node8 (None) 
```

As each job finishes, the VIBRANT "0 of 8" increments by 1 for each finished job.

I'm still waiting for everything to finish, to see if hmmscan works, but wanted to clarify this.

mshamash commented:
@replikation the -resume flag resolved the error, must have been a sporadic/one-time thing as you mention. Not sure why it happened...
I can confirm the recent commits fixed both the hmmscan database path and all database re-download bugs.

A couple points:

1. I looked into the MARVEL errors some more, since consistently only 3 of the 8 files work and the other 5 fail. Attached some output below.

This is what .command.err/.command.log look like (both have same contents):

```
Error: Sequence file ERR576943_raw_assembly.fasta_contigs/results/prokka/NODE_83_length_2127_cov_1_452220/prokka_results_NODE_83_length_2127_cov_1_452220.faa is empty or misformatted

Traceback (most recent call last):
  File "/MARVEL/marvel_bins.py", line 243, in <module>
    with open(input_folder + 'results/hmmscan/' + prefix + '_hmmscan.tbl', 'r') as hmmscan_out:
FileNotFoundError: [Errno 2] No such file or directory: 'ERR576943_raw_assembly.fasta_contigs/results/hmmscan/NODE_83_length_2127_cov_1_452220_hmmscan.tbl'
```

and .command.sh:

```
#!/bin/bash -ue
# Marvel
marvel_bins.py -i ERR576943_raw_assembly.fasta_contigs -t 24 > results_${PWD##*/}.txt
```


2. The cromomap step seems to consistently fails 50% of the time as well:

```

[0c/77d626] process > phage_annotation_MSF:prodigal (5)                                           [100%] 8 of 8 ✔
[a2/24be70] process > phage_annotation_MSF:hmmscan (7)                                            [100%] 8 of 8 ✔
[bc/eee33c] process > phage_annotation_MSF:chromomap_parser (8)                                   [100%] 8 of 8 ✔
[fc/63dbfc] process > phage_annotation_MSF:chromomap (16)                                         [100%] 34 of 34, failed: 18, retries: 18 ✔
[fe/0cc84d] process > checkV_wf:checkV (4)                                                        [100%] 8 of 8 ✔
[b4/e65d87] process > phage_tax_classification:split_multi_fasta (4)                              [100%] 8 of 8 ✔
[61/137962] process > phage_tax_classification:sourmash_for_tax (1)                               [100%] 8 of 8 ✔
```

3. Finally, the folder names for results still include ".fasta", see image below.
<img width="252" alt="output" src="https://user-images.githubusercontent.com/8537248/88095706-84e4a800-cb63-11ea-9792-495714d0d419.png">

replikation commented:
hi this looks good. So i can merge the PR into master, thank you very much so everything works. thank you very much

* regarding chromomap: works as intended: the chromomap code depending on the results. e.g. if one approach fails WtP trys to render it another way (with different settings) so we have like 4 approaches - if one fails WtP trys another one. you get 2 maps per input so 16 in case of the test profile. so everything was rendered in your case (34 process, 18 failed so 16 left that passed)

* the folder with `.fasta` will is fixed in the branch `testing--virome-flag` 

* regarding marvel if it tells "FileNotFoundError: [Errno 2] " it means it didn't produce an output` usually because nothing was identified, but be we keep an eye on this of course

* regarding the tar error, if this happens again please open an issue and I will implement a possible solution, but the error is more related to the cluster and not WtP

mshamash commented:
Great! Happy to help! Glad that everything is working. I didn't retry downloading the singularity images yet...but I think the fact that you set it to try downloading again if it fails would help users out.
